docs/extensions/api.md

```diff
@@ -2,425 +2,563 @@ title:      Extensions API
+#### Usages
 
-The new processor is very similar to legacy with two major distinctions.
+Some block processors in the Markdown source tree include:
 
-1. Patterns no longer need to match the entire block, so patterns no longer
-    start with `r'^(.*?)'` and end with `r'(.*?)!'`. This was a huge
-    performance sink and this requirement has been removed. The returned match
-    object will only contain what is explicitly matched in the pattern, and
-    extension pattern groups now start with `m.group(1)`.
+| Class  | Kind |  Description |
+| ----------------------------|-----------|--------------------------------------------------
+| [`HashHeaderProcessor`][b1] | built-in  |  Title hashes (`#`), which may split blocks |
+| [`HRProcessor`][b2]   | built-in  |  Horizontal lines, e.g., `---` |
+| [`OListProcessor`][b3] | built-in  |  Ordered lists; complex and using `state` |
+| [`Admonition`][b4] | extension |  Render each [Admonition][] in a new `div`
 
-2. The `handleMatch` method now takes an additional input called `data`,
-    which is the entire block under analysis, not just what is matched with
-    the specified pattern. The method also returns the element *and* the index
-    boundaries relative to `data` that the return element is replacing
-    (usually `m.start(0)` and `m.end(0)`).  If the boundaries are returned as
-    `None`, it is assumed that the match did not take place, and nothing will
-    be altered in `data`.
+[b1]: https://github.com/Python-Markdown/markdown/blob/master/markdown/blockprocessors.py#L441
+[b2]: https://github.com/Python-Markdown/markdown/blob/master/markdown/blockprocessors.py#L495
+[b3]: https://github.com/Python-Markdown/markdown/blob/master/markdown/blockprocessors.py#L316
+[Admonition]: https://python-markdown.github.io/extensions/admonition/
+[b4]: https://github.com/Python-Markdown/markdown/blob/master/markdown/extensions/admonition.py#L36
 
-If all you need is the same functionality as the legacy processor, you can do
-as shown below. Most of the time, simple regular expression processing is all
-you'll need.
+### Tree processors {: #treeprocessors }
 
-```python
-from markdown.inlinepatterns import InlineProcessor
-import xml.etree.ElementTree as etree
+Tree processors manipulate the tree created by block processors.  They can even create an entirely new 
+ElementTree. This is an excellent place for creating summaries, adding collected references, or last
+minute adjustments.
 
-# an oversimplified regex
-MYPATTERN = r'\*([^*]+)\*'
+A tree processor inherits from `markdown.treeprocessors.Treeprocessor`. Note it is `Treeprocessor` not `TreeProcessor`.
+A tree processor must implement a `run` method which takes 
+a single argument `root`, the current `ElementTree`.  It may either return `None`, probably modifying `root`, or an
+entirely new `ElementTree` to replace the current tree.  For efficiency, it is preferred to return `None` and 
+modify the `root` in place.
```

Comment:
It appears that this needs to be reworked as per #949. Interestingly, all references to `ElementTree` have been put into code spans here. Previously they were not because they are not referring to the `ElementTree` class, but rather any object of the ElementTree library except for the `ElementTree` object (`Element`, `Comment`, etc.).

Reply:
I'll look at this; didn't know about #949.  I'll check consistency; usually I try to refer to just 'the tree' for the current, complete `ElementTree` object of the entire Markdown document.   Let me whack at it and get back to you.